docs/programmers_guide/witness_formal_spec.md

```diff
@@ -2,72 +2,179 @@
+
+Here is how the execution code might look like in Go for building a single trie.
+
+```go
+
+witness := GetInitialWitness()
+rules := GetSubstitutionRules()
+numberOfRulesApplied := 1 // initial state
+
+for rulesApplied {
+    witness, numberOfRulesApplied := ApplyRules(witness, rules)
 }
 
-helper2 (hash1, hash2) {
-    return KECCAK(hash1 + hash2)
+if len(witness) == 1 {
+    trie.root = witness[0]
+} else {
+    panic("witness execution failed")
 }
+
 ```
 
-The specification for a substitution rule
+And here is an example of the execution process (we will use the set of rules
+form the **Substitution Rules** section of this document):
+
+Step 1. Witness: `(HASH h1 HASH h2 BRANCH 0b101 HASH h3 BRANCH 0b11)`
+
+Step 2. Applying `HASH` substitution rules.
+Witness: `(HashNode{h1} HashNode{h2} BRANCH 0b101 HashNode{h3} BRANCH 0b11)`
+
+Step 3. Applying `BRANCH` substitution rules (only once, because `BRANCH 0b11`
+doesn't pass its `GUARD` statements just yet).
+Witness: `(BranchNode{0: HashNode{h1} 2:HashNode{h2}} HashNode{h3} BRANCH 0b11)`
+
+Step 4: Applying `BRANCH` substitution rules again.
+Witness: `(BranchNode{0: BranchNode{0: HashNode{h1} 2:HashNode{h2}} 1:HashNode{h3}})`
+
+Step 5: No more rules are applicable, the execution ends successfully.
+
+
+## End Criteria
+
+The execution ends when there are no substitution rules applicable for this
+witness.
+
+### Building a single trie from the witness
+
+If we are building a single trie from the witness, then the only SUCCESS
+executon is when the following are true:
+
+- The execution state MUST match the End Criteria
+- There MUST be only one item left in the witness
+- This item MUST be of the type `Node`
+    
+In that case, this last item will be the root of the built trie.
+
+Every other end state is considered a FAILURE.
+
+
+### Building a Forest 
+
+We also can build a forest of tries with this approach, by adding a new
+Instruction `NEW_TRIE` and adjusting the success criteria a bit:
+
+- The execution state MUST match the End Criteria;
+- The items that are left in the witness MUST follow this pattern: `(Node
+    NEW_TRIE ... Node)`
+- Each `Node` element will be a root of a trie.
+
+Every other end state is considered a FAILURE.
 
-```
-[GUARD <CONDITION> ...]
 
-[ STACK(<node-var-name>, <hash-var-name>), ... ] <INSTRUCTION>[(<params>)] |=>
-STACK(node-value, hash-value), ...
+## Instructions & Parameters
+
+A single instruction is consists of substitution rules and parameters.
+
+Each instruction MAY have one or more parameters.
+The parameters values MUST be situated in the witness.
+The parameter values MUST NOT be taken from the stack.
+
+That makes it different from the helper function parameters that MAY come from the stack or MAY come from the witness.
+
+
+## Substitution rules
+
+A substitution rule consists of 2 parts: the criteria (on the left of the `|=>` sign) and the result (on the right of the `|=>` sign).
```

Comment:
I would separate pattern, guard, and the result as separate things. And we need to mention that the "pattern" part may contain variables that get bound when the pattern matched. Then, these bound variables can be used in the guards and in the result.